runtime/compiler/optimizer/J9TransformUtil.cpp

```diff
@@ -1982,3 +1982,63 @@ void J9::TransformUtil::prohibitOSROverRange(TR::Compilation* comp, TR::TreeTop*
       }
    }
 
+static void prohibitGuardedStaticFinalFieldFoldingOnNodeAndChildren(TR::Compilation* comp, TR::Node* node, TR::NodeChecklist *visited)
+   {
+   if (visited->contains(node))
+      return;
+
+   visited->add(node);
+
+   int32_t numChildren = node->getNumChildren();
+   if (numChildren == 0 &&
+       node->getOpCode().isLoadVarDirect() &&
+       node->isLoadOfStaticFinalField())
+      {
+      traceMsg(comp, "prohibit folding on n%dn\n", node->getGlobalIndex());
+      node->getByteCodeInfo().setDoNotProfile(true);
+      return;
+      }
+
+   for (int i=0; i<numChildren; i++)
+       {
+       TR::Node* child = node->getChild(i);
+       prohibitGuardedStaticFinalFieldFoldingOnNodeAndChildren(comp, child, visited);
+       }
+   }
+
+/** \brief
+ *     Prohibit guarded static final field folding over range between start and end inclusive.
+ *
+ *     Walk region and ensure all static final field loads has doNotProfile set on their bci.
+ *
+ *  \param comp
+ *     The compilation object.
+ *
+ *  \param start
+ *     The tree marking the start of the range.
+ *
+ *  \param end
+ *     The tree marking the end of the range.
+ */
+void J9::TransformUtil::prohibitGuardedStaticFinalFieldFoldingOverRange(TR::Compilation* comp, TR::TreeTop* start, TR::TreeTop* end)
+   {
+   TR_ASSERT(start->getEnclosingBlock() == end->getEnclosingBlock(), "Does not support range across blocks");
+
+   traceMsg(comp, "prohibit final field folding over n%dn - n%dn \n", start->getNode()->getGlobalIndex(), end->getNode()->getGlobalIndex());
+   TR::TreeTop* ttAfterEnd = end->getNextTreeTop();
+
+   TR::NodeChecklist visited(comp);
+
+   if (comp->getOption(TR_EnableOSR) &&
+       comp->isOSRTransitionTarget(TR::postExecutionOSR) &&
+       comp->getOSRMode() == TR::voluntaryOSR)
```

Comment:
prohibition only works in certain modes - we should probably FATAL_ASSERT if it isn't supported or have some kind of return code. We shouldn't just silently fail.

Reply:
Added an assertion and StringPeepHoles now checks the mode before prohibition.